src/wallet/wallet-generation.ts

```diff
@@ -0,0 +1,198 @@
+import axios from 'axios'
+
+import {RippleAPI} from '..'
+import {errors} from '../common'
+import {GeneratedAddress} from '../offline/generate-address'
+import {isValidAddress} from '../common/schema-validator'
+import {reject} from 'lodash'
+
+export interface FaucetWallet {
+  account: GeneratedAddress
+  amount: number
+  balance: number
+}
+
+export enum FaucetNetwork {
+  Testnet = 'https://faucet.altnet.rippletest.net/accounts',
+  Devnet = 'https://faucet.devnet.rippletest.net/accounts'
+}
+
+const INTERVAL_SECONDS = 1 // Interval to check an account balance
+const MAX_ATTEMPS = 20 // Maximum attemptsto retrieve a balance
+
+/**
+ * Generates a random wallet with some amount of XRP (usually 1000 XRP).
+ *
+ * @param wallet - An existing XRPL wallet to fund, if undefined, a new wallet will be created.
+ * @returns - A Wallet on the Testnet or Devnet that contains some amount of XRP.
+ */
+async function generateFaucetWallet(
+  this: RippleAPI,
+  wallet?: string
+): Promise<FaucetWallet | void> {
+  let body = {}
+  let startingBalance = 0
+  let address: string
+
+  // If the user provides an existing wallet to fund
+  if (wallet && isValidAddress(wallet)) {
+    // Create the POST request body
+    body = {
+      destination: wallet
+    }
+    // Retrieve the existing account balance
+    try {
+      const accountToFundBalance = await getWalletBalanceByCurrency(
+        this,
+        wallet
+      )
+      if (accountToFundBalance) {
+        startingBalance = +accountToFundBalance
+      } else {
+        startingBalance = 0
+      }
+    } catch (err) {
+      reject(
+        new errors.XRPLFaucetError(
+          `Unable to retrieve the balance of your account ${wallet}`
+        )
+      )
+    }
+    // And its address
+    address = wallet
+  }
+
+  // Based on the connection URL, retrieve the correct faucet URL
+  const faucetNetwork = getFaucetUrl(this)
+
+  return new Promise((resolve, reject) => {
+    axios
+      .post<FaucetWallet>(faucetNetwork, body)
+      .then(async (response) => {
+        address = response.data.account.classicAddress
```

Comment:
this shouldn't be using the same variable as `address` above - it's confusing

Reply:
☝️

Reply:
Agree, I will rename it to `classicAddress`.